crux_macros/src/effect.rs

```diff
@@ -66,12 +66,46 @@ impl ToTokens for EffectStructReceiver {
         let mut with_context_fields = Vec::new();
         let mut ffi_variants = Vec::new();
         let mut match_arms = Vec::new();
+        let mut try_from_impls = Vec::new();
 
         for (field_name, (capability, variant, event)) in fields.iter() {
             variants.push(quote! { #variant(::crux_core::Request<<#capability<#event> as ::crux_core::capability::Capability<#event>>::Operation>) });
             with_context_fields.push(quote! { #field_name: #capability::new(context.specialize(#effect_name::#variant)) });
             ffi_variants.push(quote! { #variant(<#capability<#event> as ::crux_core::capability::Capability<#event>>::Operation) });
             match_arms.push(quote! { #effect_name::#variant(request) => request.serialize(#ffi_effect_name::#variant) });
+
+            let filter_predicate_ident = format_ident!("filter_{}", field_name);
+            let map_predicate_ident = format_ident!("map_{}", field_name);
+            try_from_impls.push(quote! {
+                impl TryFrom<&#effect_name> for <#capability<#event> as ::crux_core::capability::Capability<#event>>::Operation {
+                    type Error = ();
+
+                    fn try_from(value: &#effect_name) -> Result<Self, Self::Error> {
+                        if let #effect_name::#variant(request) = value {
+                            Ok(request.operation.clone())
+                        } else {
+                            Err(())
+                        }
+                    }
+                }
+
+                impl #effect_name {
+                    pub fn #filter_predicate_ident(effect: &#effect_name) -> bool {
+                        if let #effect_name::#variant(_) = effect {
+                            true
+                        } else {
+                            false
+                        }
+                    }
+                    pub fn #map_predicate_ident(effect: #effect_name) -> Option<crux_core::Request<<#capability<#event> as ::crux_core::capability::Capability<#event>>::Operation>> {
+                        if let #effect_name::#variant(request) = effect {
+                            Some(request)
+                        } else {
+                            None
+                        }
+                    }
+                }
+            });
```

Comment:
Nice, yeah - this is the kind of thing I was thinking.

Wondering about the names though:

- `filter_{}` could maybe be `is_{}`?
- `map_x` is _almost_ a conversion method, so wonder if it should follow [the conventions](https://rust-lang.github.io/api-guidelines/naming.html#ad-hoc-conversions-follow-as_-to_-into_-conventions-c-conv) and be named `into_x`?